### PR TITLE
Changing the implementation of MultipleApplicationsSource to be parallel

### PR DESCRIPTION
**Pull request by madoar**
This PR changes the implementation of MultipleApplicationsSource to be parallel through the use of the Java 8 Streaming API, see also #650. 
In addition this PR extracts the merging functionality of two `CategoryDTOs` from the `TeeApplicationsSource` class to a new interface called `MergeableApplicationsSource`, which enables this functionally to be directly used in `MultipleApplicationsSource`.

### CODE REVIEW

**Comment by plata** [comment]
ok to test

**Comment by madoar** [comment]
In the context of this PR, we can think about moving the for loop over multiple (two or more) application sources to the new `MergeableApplicationsSource` interface too, because both `MultipleApplicationsSource` and `TeeApplicationsSource` use this code.

**Comment by plata** [comment]
If we have duplicate code, we should put that in a common class.

File: phoenicis-apps/src/main/java/org/phoenicis/apps/MergeableApplicationsSource.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+import java.util.SortedMap;
+import java.util.TreeMap;
+import java.util.function.Function;
+
+import org.phoenicis.apps.dto.ApplicationDTO;
+import org.phoenicis.apps.dto.CategoryDTO;
+import org.phoenicis.apps.dto.ResourceDTO;
+import org.phoenicis.apps.dto.ScriptDTO;
+
+public interface MergeableApplicationsSource extends ApplicationsSource {
```
Comment (qparis) [inline phoenicis-apps/src/main/java/org/phoenicis/apps/MergeableApplicationsSource.java]: It looks like an Abstract class more than an interface

File: phoenicis-apps/src/main/java/org/phoenicis/apps/MergeableApplicationsSource.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+	public default CategoryDTO mergeCategories(CategoryDTO leftCategory, CategoryDTO rightCategory) {
```
Comment (qparis) [inline phoenicis-apps/src/main/java/org/phoenicis/apps/MergeableApplicationsSource.java]: You need to reformat your code here

File: phoenicis-apps/src/main/java/org/phoenicis/apps/MultipleApplicationsSource.java [hunk 3]
```diff
@@ -18,30 +18,71 @@
...
+		 * This step is needed because we need a mapping between the CategoryDTO
+		 * list and its application source, to preserve the order in the
+		 * reduction step
+		 */
+		final Map<ApplicationsSource, List<CategoryDTO>> categoriesMap = this.applicationsSources.parallelStream()
+				.collect(
+						Collectors.toConcurrentMap(source -> source, ApplicationsSource::fetchInstallableApplications));
+
+		if (applicationsSources.isEmpty())
+			return Collections.emptyList();
```
Comment (qparis) [inline phoenicis-apps/src/main/java/org/phoenicis/apps/MultipleApplicationsSource.java]: Please put braces here

File: phoenicis-apps/src/main/java/org/phoenicis/apps/MergeableApplicationsSource.java (hunk 2 above)
Comment (madoar) [inline phoenicis-apps/src/main/java/org/phoenicis/apps/MergeableApplicationsSource.java]: What do you mean with "reformat" here? Do you mean I'm using the wrong formatter settings? Or that this shouldn't be an interface but a class?

**Comment by madoar** [comment]
The parallelization happens in line 48 with the method call `parallelStream()` this makes java call `ApplicationsSource::fetchInstallableApplications`, which can take a lot of time, for every application source in parallel.
The merging step is still done sequentially in the for loop afterwards.

**Pull request merged**